fiber/fiber_wasmfx.c

```diff
@@ -0,0 +1,123 @@
+// An implementation of the fiber.h interface using wasmfx continuations
+#include <stdlib.h>
+#include <stdint.h>
+
+#include "fiber.h"
+#include "wasm.h"
+
+// Type for indices into the continuation table `$conts` on the wasm side. In
+// this implementation, we consider `fiber_t` (which is a pointer to an
+// incomplete type) to be equivalent to `cont_table_index_t` and freely convert
+// between the two.
+typedef uintptr_t cont_table_index_t;
+
+// Initial size of the `$conts` table. Keep this value in sync with the
+// corresponding (table ...) definition.
+static const size_t INITIAL_TABLE_CAPACITY = 1024;
+
+// The current capacity of the `$conts` table.
+static size_t cont_table_capacity = INITIAL_TABLE_CAPACITY;
+// Number of entries at the end of `$conts` table that we have never used so
+// far.
+// Invariant:
+// `cont_table_unused_size` + `free_list_size` <= `cont_table_capacity`
+static size_t cont_table_unused_size = INITIAL_TABLE_CAPACITY;
+
+// This is a stack of indices into `$conts` that we have previously used, but
+// subsequently freed. Allocated as part of `fiber_setup`. Invariant: Once
+// allocated, the capacity of the `free_list` (i.e., the number of `size_t`
+// values we allocate memory for) is the same as `cont_table_capacity`.
+static size_t* free_list = NULL;
+// Number of entries in `free_list`.
+// Invariant: free_list_size <= `cont_table_capacity`.
+static size_t free_list_size = 0;
+
+
+extern
+__wasm_import("fiber_wasmfx_imports", "wasmfx_grow_cont_table")
+void wasmfx_grow_cont_table(size_t);
+
+extern
+__wasm_import("fiber_wasmfx_imports", "wasmfx_indexed_cont_new")
+void wasmfx_indexed_cont_new(fiber_entry_point_t, cont_table_index_t);
+
+extern
+__wasm_import("fiber_wasmfx_imports", "wasmfx_indexed_resume")
+void* wasmfx_indexed_resume(size_t fiber_index, void *arg, fiber_result_t *result);
+
+extern
+__wasm_import("fiber_wasmfx_imports", "wasmfx_suspend")
+void* wasmfx_suspend(void *arg);
+
+
+ cont_table_index_t wasmfx_acquire_table_index() {
+  uintptr_t table_index;
+  if (cont_table_unused_size > 0) {
+    // There is an entry in the continuation table that has not been used so far.
+    table_index = cont_table_capacity - cont_table_unused_size;
+    cont_table_unused_size--;
+  } else if (free_list_size > 0) {
+      // We can pop an element from the free list stack.
+      table_index = free_list[free_list_size - 1];
```

Comment:
We should experiment with different policies here. I suppose for large freelists it may be bad for locally to read from the end of the list instead of the start. Similarly, the current policy seem to prefer unused cont indices, it may be better to prefer to use the freelist entries first, if those entries are at the low end of the table -- again because of locality.

Reply:
These are exactly the kinds of things I've tried.

Reply:
What did you find?